### ISSUE TITLE
Get key's value return null

### ISSUE THREAD

**Issue by zhazhuanling** [redis/hiredis#865]
On Centos7.8, compiling example.c,  after connecting the redis and set the key and value, but the problem is that:
     it can't get the value but with null. Here are the print message.

PING: (null)
SET: (null)
SET (binary API): (null)
GET foo: (null)
INCR counter: 0
INCR counter: 0

I want to get key's value.  Can you tell me if you have any information?

**Comment by michael-grunder** [comment]
It's hard to say what issue you're running into.  Are you sure you're running Redis on the machine?

It seems unlikely that the example is completely broken since we'd have lots of reports.  Can you replicate this in a container?

**Comment by zhazhuanling** [comment]
@michael-grunder Redis is running, and the machine is generated by docker.  After running the example,  the key's value in Redis has been updated(but get failed by the c++ program). And it's failed while running in another machine. Fortunately the following code get success with same env! But I don't know why.   |||-_-

redis.h
```
#ifndef _REDIS_H_
#define _REDIS_H_

#include <iostream>
#include <string.h>
#include <string>
#include <stdio.h>

#include <hiredis/hiredis.h>

class Redis
{
public:

    Redis(){}

    ~Redis()
    {
        this->_connect = NULL;
        this->_reply = NULL;                
    }

    bool connect(std::string host, int port)
    {
        this->_connect = redisConnect(host.c_str(), port);
        if(this->_connect != NULL && this->_connect->err)
        {
            printf("connect error: %s\n", this->_connect->errstr);
            return 0;
        }
        return 1;
    }

    std::string get(std::string key)
    {
        this->_reply = (redisReply*)redisCommand(this->_connect, "GET %s", key.c_str());
        std::string str = this->_reply->str;
        freeReplyObject(this->_reply);
        return str;
    }

    void set(std::string key, std::string value)
    {
        redisCommand(this->_connect, "SET %s %s", key.c_str(), value.c_str());
    }

private:

    redisContext* _connect;
    redisReply* _reply;
                
};

#endif  //_REDIS_H_
```
redis.cpp
```
#include "redis.h"
int main()
{
    Redis *r = new Redis();
    if(!r->connect("127.0.0.1", 6379))
    {
        printf("connect error!\n");
        return 0;
    }
    r->set("name", "Andy");
    printf("Get the name is %s\n", r->get("name").c_str());
    delete r;
    return 0;
}
```

**Comment by michael-grunder** [comment]
I'm glad you got it working.  I'm going to close the issue since it's unlikely there is a bug in Hiredis but if you're still having trouble feel free to post some code that isn't working and I'll take a look.

**Comment by arnaud-soulard** [comment]
I have a very similar issue.
I am on CentOS Linux release 7.7.1908.
I am not able to get a anything else than NULL in reply->str.
The same piece of code works perfectly with hiredis version 0.12.1. (I get the expected value in reply->str.)
It does not work with hiredis version 773d6ea8a73de57279a8733aa884bc5710816779.
Hiredis 1.0.0 release version does not compile ("SET(CMAKE_C_STANDARD 99)" was missing)


The piece of code looks like this:
redisContext *context = redisConnectUnix(...);
...
redisReply* reply = (redisReply*)redisCommand(context, "INFO");
//////////reply->type == REDIS_REPLY_STRING
reply->str is always NULL


If you have an idea about what I am missing, I really would appreciate!

**Comment by michael-grunder** [comment]
> Hiredis 1.0.0 release version does not compile ("SET(CMAKE_C_STANDARD 99)" was missing)

You might try building and installing Hiredis by using the makefile (instead of CMake).

I don't have enough information from what you posted to hazard a guess as to what might be going wrong.  Perhaps you can post a complete program that I can look at.

Are you sure the permissions are correct on the unix socket?

**Comment by arnaud-soulard** [comment]
I am using cmake because my project is using cmake as well.
The project is using old version of hiredis (0.12.1) installed through CentOS, and I want to use the last hiredis version (1.0.0) on the part of the project I am working on.

There is a "find_package(Hiredis REQUIRED)" that finds the old version (0.12.1).

I just included hiredis sources (1.0.0) like this:
add_outdirectory (${PROJECT_ROOT_DIR} hiredis)
It gets compiled with the project.

I tried to define HIREDIS_FOUND/HIREDIS_INCLUDE_DIRS/HIREDIS_LIBRARIES (adding SDK_ to not get confused with old version of hiredis found on the system)


set(SDK_HIREDIS_FOUND TRUE CACHE BOOL "Whether hiredis is found on the system or not")
set(SDK_HIREDIS_INCLUDE_DIRS ${hiredis_INCLUDE_DIR} CACHE PATH "hiredis include directory")
set(SDK_HIREDIS_LIBRARIES hiredis CACHE INTERNAL "hiredis libraries")



Here is the piece of code I am using:


redisContext *c;
redisReply *reply;

c = redisConnectUnix("/path/to/unix.sock");

/* INFO */
reply = static_cast<redisReply *>(redisCommand(c,"INFO"));
printf("INFO: %s\n", reply->str); /////////////////////////////////////////////////reply->str is NULL
freeReplyObject(reply);

/* Set a key */
reply = static_cast<redisReply *>(redisCommand(c,"SET %s %s", "foo", "hello world"));
printf("SET: %s\n", reply->str); /////////////////////////////////////////////////reply->str is NULL, but the key foo has been set (I checked with a redis-cli client)
freeReplyObject(reply);

/* Try a GET */
reply = static_cast<redisReply *>(redisCommand(c,"GET foo"));
printf("GET foo: %s\n", reply->str); /////////////////////////////////////////////////reply->str is NULL
freeReplyObject(reply);

It seems that hiredis works for me to set a key, but not to get its value. I suppose the permission are OK, if not, I think I would not be able to use hiredis at all.
Do you have an idea?

**Comment by michael-grunder** [comment]
Is it possible that your program is linking with the old version of hiredis but compiling with the newer version?  That kind of ABI problem could account for this kind of weird behavior.

Another thing you could try is using a C++ library instead (e.g. [hireids-vip](https://github.com/vipshop/hiredis-vip) or [redis-plus-plus](https://github.com/sewenew/redis-plus-plus).  It might be more ideal anyway since your project is in C++.

If you create a container with a small example project that demonstrates the bug I'd be happy to track down what's happening.  I can't say without code I can run.
